orchestrator should fix tasks before start

dongluochen opened the pull request:
Fix #1812. 

Global orchestrator considers a `rejected` task not running and adds a new task to it. But this should be handled by task restart logic.  
 
```
gtest.0.tvi9nt4n7ghn  busybox:latest  ip-172-19-241-146          Ready          Rejected 1 second ago            "failed to allocate gateway (1…"
```

Signed-off-by: Dong Chen <dongluo.chen@docker.com>

aaronlehmann commented:
I think this is right. I need to take a closer look since it touches a few different code paths. But it seems correct that service-level reconciliation should only look at desired state.

aaronlehmann commented:
Is there harm in this beyond incorrect counts temporarily showing up in the CLI? We might want to push this to 1.13.1 or 1.14.

aaronlehmann commented:
I'm wondering what happens if there is already a task that needs to be restarted when the global orchestrator starts up. There will be no event triggering a restart for that task, so I think it would not get restarted.

I thought the replicated orchestrator handled this correctly, but looking at the code, I'm not seeing it. In `initTasks`, it's checking for running tasks on invalid nodes, but not tasks that simply failed and never had their desired state set to "shutdown". So the replicated orchestrator may have this problem already.

Unless I'm missing something, it may make sense to fix this in the replicated orchestrator, and then factor out `initTasks` and `tickTasks` so the global orchestrator can reuse them.

dongluochen commented:
> Is there harm in this beyond incorrect counts temporarily showing up in the CLI?

Since multiple tasks were in `desired state - ready, task state - rejected`, if they all get restarted successfully, there could be multiple tasks of a global service running for a short period of time, until next reconciliation loop kicks in. I think it's not a big problem. We can fix it in 1.13.1.

dongluochen commented:
> it may make sense to fix this in the replicated orchestrator, and then factor out initTasks and tickTasks so the global orchestrator can reuse them.

I think this is right. Before going into event-driven loop, orchestrators need to set things right.

aaronlehmann commented:
Hi @dongluochen - are you still working on this? I don't know if the fix will make it into 1.13, but it would be good to have for 1.13.1, at least.

dongluochen commented:
@aaronlehmann I'm working on it. Will also update replicated orchestrator.

aaronlehmann commented:
ping - I think 1.13.1 is getting to the -rc phase.

aaronlehmann commented on manager/orchestrator/replicated/replicated_test.go:
```diff
@@ -424,3 +425,133 @@ func TestReplicatedScaleDown(t *testing.T) {
+		select {
+		case <-watch:
+		case <-time.After(time.Second):
+			done = true
```
Let's count watch events instead of using wall clock time.

dongluochen replied:
There are the events for slot 1. `initTask()` tries to restart the task, moving it DesiredState from `READY` to `RUNNING`, then `tick()` would move DesiredState to `SHUTDOWN`, and start a new task to replace it. So it's doing the right thing. There is a lot of implementation details in the event sequence, and they may change when implementation changes. That's why I try to check the end result instead of the events. 

```
event: {&Task{ID:task1,Meta:Meta{Version:Version{Index:0,},CreatedAt:(timestamp: nil google_protobuf.Timestamp),UpdatedAt:(timestamp: nil google_protobuf.Timestamp),},Spec:TaskSpec{Runtime:<nil>,Resources:nil,Restart:nil,Placement:nil,LogDriver:nil,Networks:[],ForceUpdate:0,},ServiceID:id1,Slot:1,NodeID:node1,Annotations:Annotations{Name:,Labels:map[string]string{},},ServiceAnnotations:Annotations{Name:task1,Labels:map[string]string{},},Status:TaskStatus{Timestamp:(timestamp: nil google_protobuf.Timestamp),State:REJECTED,Message:,Err:,RuntimeStatus:<nil>,PortStatus:nil,},DesiredState:RUNNING,Networks:[],Endpoint:nil,LogDriver:nil,} []}

event: {&Task{ID:task1,Meta:Meta{Version:Version{Index:0,},CreatedAt:(timestamp: nil google_protobuf.Timestamp),UpdatedAt:(timestamp: nil google_protobuf.Timestamp),},Spec:TaskSpec{Runtime:<nil>,Resources:nil,Restart:nil,Placement:nil,LogDriver:nil,Networks:[],ForceUpdate:0,},ServiceID:id1,Slot:1,NodeID:node1,Annotations:Annotations{Name:,Labels:map[string]string{},},ServiceAnnotations:Annotations{Name:task1,Labels:map[string]string{},},Status:TaskStatus{Timestamp:(timestamp: nil google_protobuf.Timestamp),State:REJECTED,Message:,Err:,RuntimeStatus:<nil>,PortStatus:nil,},DesiredState:SHUTDOWN,Networks:[],Endpoint:nil,LogDriver:nil,} []}

event: {&Task{ID:ainr2bk0mkxccn27i89lg7tcu,Meta:Meta{Version:Version{Index:0,},CreatedAt:(timestamp: nil google_protobuf.Timestamp),UpdatedAt:(timestamp: nil google_protobuf.Timestamp),},Spec:TaskSpec{Runtime:<nil>,Resources:nil,Restart:nil,Placement:nil,LogDriver:nil,Networks:[],ForceUpdate:0,},ServiceID:id1,Slot:1,NodeID:,Annotations:Annotations{Name:,Labels:map[string]string{},},ServiceAnnotations:Annotations{Name:name1,Labels:map[string]string{},},Status:TaskStatus{Timestamp:2017-02-01T00:07:11.897776845Z,State:NEW,Message:created,Err:,RuntimeStatus:<nil>,PortStatus:nil,},DesiredState:READY,Networks:[],Endpoint:&Endpoint{Spec:nil,Ports:[],VirtualIPs:[],},LogDriver:nil,} []}
```

aaronlehmann commented on manager/orchestrator/replicated/replicated_test.go:
```diff
@@ -424,3 +425,133 @@ func TestReplicatedScaleDown(t *testing.T) {
+				assert.FailNow(t, "unexpected task DesiredState")
+			}
+		}
+		require.Len(t, active, 2)
```
I think this is not checking the right things. The two tasks that were created above have desired state "ready", so they would count as the two "active" tasks. There is no check that the orchestrator actually advanced their desired state to "shutdown" and replaced them. From my reading of the current code, it looks like the replicated orchestrator is not doing this at startup, even though it should.

dongluochen replied:
See above.

aaronlehmann commented on manager/orchestrator/replicated/replicated_test.go:
```diff
@@ -424,3 +425,133 @@ func TestReplicatedScaleDown(t *testing.T) {
+			case task.DesiredState <= api.TaskStateRunning:
+				active = append(active, task)
+			case task.DesiredState == api.TaskStateShutdown:
+				assert.True(t, task.Status.State == api.TaskStateRejected)
```
I suspect this is never reached in the test right now. This case should also add the tasks to a slice that's checked later, instead of asserting.

dongluochen replied:
You can find the event above. This tasks are moved to `SHUTDOWN`. I can update the test to follow the event sequence, if you think that's better.

aaronlehmann replied:
Sorry, I'm still not seeing how the test confirms this. It looks like the `for !done {` loop is completely ignoring all the watch activity, and is equivalent to just `time.Sleep(time.Second)`.

dongluochen commented:
@aaronlehmann I refactored `initTasks` as you suggested. Added `restarter` interface to remove dependency cycle. `restart` is tight coupled with `orchestrator` so the interface looks a little messy.

aaronlehmann commented on manager/orchestrator/task.go:
```diff
@@ -65,3 +75,119 @@ func IsTaskDirty(s *api.Service, t *api.Task) bool {
 		(t.Endpoint != nil && !reflect.DeepEqual(s.Spec.Endpoint, t.Endpoint.Spec))
 }
+
+// InvalidNode is true if the node is nil, down, or Drain
```
s/Drain/drained/

aaronlehmann commented on manager/orchestrator/task.go:
```diff
@@ -65,3 +75,119 @@ func IsTaskDirty(s *api.Service, t *api.Task) bool {
+
+type isRelatedService func(*api.Service) bool
+
+// InitTasks fix tasks in store before orchestrator run. Previous manager might haven't finished processing their updates.
```
`InitTasks fixes tasks in the store before the orchestrator runs. The previous leader might not have finished processing their updates and left them in an inconsistent state.`

aaronlehmann commented on manager/orchestrator/task.go:
```diff
@@ -65,3 +75,119 @@ func IsTaskDirty(s *api.Service, t *api.Task) bool {
+				if InvalidNode(n) && t.DesiredState <= api.TaskStateRunning {
+					t.DesiredState = api.TaskStateShutdown
+					err = batch.Update(func(tx store.Tx) error {
+						return store.UpdateTask(tx, t)
```
Interesting approach. I'm not sure this is the best way. In the future we might filter out updates where nothing changed. I wonder if it's better for this function to return a slice of tasks that need to be reconciled.

aaronlehmann commented on manager/orchestrator/task.go:
```diff
@@ -65,3 +75,119 @@ func IsTaskDirty(s *api.Service, t *api.Task) bool {
+			// can look at it
+			if t.DesiredState <= api.TaskStateRunning && t.Status.State > api.TaskStateRunning {
+				err := batch.Update(func(tx store.Tx) error {
+					return store.UpdateTask(tx, t)
```
Interesting approach. I'm not sure this is the best way. In the future we might filter out updates where nothing changed. I wonder if it's better for this function to return a slice of tasks that need to be reconciled.

aaronlehmann commented on manager/orchestrator/task.go:
```diff
@@ -65,3 +75,119 @@ func IsTaskDirty(s *api.Service, t *api.Task) bool {
+					t.DesiredState = api.TaskStateShutdown
+					err = batch.Update(func(tx store.Tx) error {
+						return store.UpdateTask(tx, t)
+					})
```
I'm not sure this will do the right thing. In the replicated case, the task may need to be restarted, but I believe that if the orchestrator sees a task move its desired state to `TaskStateShutdown`, it will assume it has already been restarted.

dongluochen replied:
I was thinking about future orchestrator like `Batch`. It shouldn't restart the task.

If restart behavior is related to orchestrator types, it's hard to get right.

aaronlehmann replied:
We already support restarts in this function as part of the `restartDelay` logic below.

Right now I believe the restart supervisor looks at what kind of service the task is part of. In the future we might want to refactor this to use a different restarter implementation for each service type.

But regardless of any refactoring, I think it's important that in this case we don't shut down the task but never start a replacement. That would be incorrect behavior.

aaronlehmann commented on manager/orchestrator/global/global_test.go:
```diff
@@ -267,3 +267,197 @@ func deleteTask(t *testing.T, s *store.MemoryStore, task *api.Task) {
+	observedTask1 := testutils.WatchTaskUpdate(t, watch)
+	assert.True(t, observedTask1.ID == "task1" || observedTask1.ID == "task2")
+	assert.Equal(t, observedTask1.Status.State, api.TaskStateRunning)
+	assert.Equal(t, observedTask1.DesiredState, api.TaskStateShutdown)
```
Maybe we should do a `View` after this and make sure we see only one task with a desired state between "ready" and "running".

dongluochen replied:
will do.

aaronlehmann commented on manager/orchestrator/replicated/replicated_test.go:
```diff
@@ -424,3 +427,324 @@ func TestReplicatedScaleDown(t *testing.T) {
+	assert.Equal(t, observedTask3.DesiredState, api.TaskStateReady)
+}
+
+func TestInitializationDelayStart(t *testing.T) {
```
It's great to have a test for this. I don't think it was covered before.

aaronlehmann commented:
I think it's worth adding another test for a preexisting task which is assigned to a node that is down, drained, or doesn't exist. From my reading of the code, I don't think this is handled correctly so far (I don't think a replacement task would be started if the restart policy calls for that).

aaronlehmann commented on manager/orchestrator/task.go:
```diff
@@ -6,9 +6,19 @@ import (
+	"golang.org/x/net/context"
 )
 
+// Restarter defines an interface to start or delay start tasks
```
s/Restarter/restarter/
s/delay start/delay starting/

aaronlehmann commented on manager/orchestrator/task.go:
```diff
@@ -6,9 +6,19 @@ import (
+type restarter interface {
+	StartNow(tx store.Tx, taskID string) error
+	DelayStart(ctx context.Context, _ store.Tx, oldTask *api.Task, newTaskID string, delay time.Duration, waitStop bool) <-chan struct{}
+}
```
As an alternative to this interface, we could introduce an `orchestrator/init` package that can import both `orchestrator` and `orchestrator/restart`.

dongluochen replied:
I feel like `initTask` should be done once before starting orchestrators. But restart may be related to specific orchestrator. If we can clarify the rules, it'd help.

aaronlehmann replied:
If you think it makes sense we can define an exported `Restarter` interface in the `orchestrator` package and use that consistently, but I'm not a big fan of the private interface as a workaround.

aaronlehmann commented on manager/orchestrator/replicated/replicated_test.go:
```diff
@@ -424,3 +427,651 @@ func TestReplicatedScaleDown(t *testing.T) {
+	// it has not been scheduled
+	assert.Equal(t, observedTask2.NodeID, "")
+	assert.Equal(t, observedTask2.Status.State, api.TaskStateNew)
+	assert.Equal(t, observedTask2.DesiredState, api.TaskStateReady)
```
`TestInitializationFailedTasks` below has a `View` that counts tasks. This test should too.

aaronlehmann commented on manager/orchestrator/replicated/replicated_test.go:
```diff
@@ -424,3 +427,651 @@ func TestReplicatedScaleDown(t *testing.T) {
+	assert.Equal(t, observedTask2.DesiredState, api.TaskStateReady)
+}
+
+func TestInitializationNodeDrain(t *testing.T) {
```
I think this is already done by `TestDrain`. But I guess having this extra test doesn't hurt.

aaronlehmann commented on manager/orchestrator/global/global.go:
```diff
@@ -171,6 +180,58 @@ func (g *Orchestrator) Run(ctx context.Context) error {
+		g.removeTask(ctx, batch, t)
+		return
+	}
+	// if the node doesn't meet service constraint, remove the task
```
This is redundant with `constraintenforcer`. Either the orchestrator should not do this at init, or `constraintenforcer` should not do it. I'd vote for keeping it in `constraintenforcer` because that seems like a logical place, and it's less of a change that way.

aaronlehmann commented on manager/orchestrator/replicated/tasks.go:
```diff
@@ -228,7 +142,49 @@ func (r *Orchestrator) handleTaskChange(ctx context.Context, t *api.Task) {
 		r.restartTasks[t.ID] = struct{}{}
+		return
+	}
+	// Ensure that the node meets scheduling constraints.
```
This is redundant with `constraintenforcer`. Either the orchestrator should not do this at init, or `constraintenforcer` should not do it. I'd vote for keeping it in `constraintenforcer` because that seems like a logical place, and it's less of a change that way.

aaronlehmann commented on manager/orchestrator/taskinit/init.go:
```diff
@@ -0,0 +1,103 @@
+					return store.DeleteTask(tx, t.ID)
+				})
+				if err != nil {
+					log.G(ctx).WithError(err).Error("failed to set task desired state to dead")
```
The log message is misleading. This code is deleting the task, not changing its desired state.

aaronlehmann commented on manager/orchestrator/taskinit/init.go:
```diff
@@ -0,0 +1,103 @@
+
+			// TODO(aluzzardi): We should NOT retrieve the service here.
+			service := store.GetService(readTx, t.ServiceID)
+			if service == nil {
```
It's a little weird that this case is being handled by both orchestrators' `CheckTasks` call. They can't step on each other because `Batch` locks the store, but it seems redundant. Maybe in the future it's a good idea to have `CheckTasks` only run once, and call into either orchestrator depending what kinds of tasks it sees. This would eliminate an extra expensive `store.FindTasks(readTx, store.All)`. I'm not sure what would call `CheckTasks` though.

aaronlehmann commented:
LGTM
